File: 04-vector-search/python/Blobs/loadAndVectorize.py
# Import necessary modules
import os
import json
import datetime
from pymongo import UpdateOne
import Collections.customers as Customers
import Collections.products as Products
import Collections.salesOrders as SalesOrders
import SearchComponents.indexes as Indexes

# Function to load and vectorize local blob data to MongoDB cluster
def loadAndVectorizeLocalBlobDataToMongoDBCluster(client, data_folder,cosmos_db_mongodb_database,batch_size,embeddings_deployment,AzureOpenAIClient, process_customers_vector, process_products_vector, process_sales_orders_vector):
    # Get list of files in the data folder
    local_blobs_files = os.listdir(data_folder)

    # Iterate over each file in the folder
    for blob_file in local_blobs_files:
        batch_number = 1

        # Process only JSON files
        if blob_file.find(".json") >= 0:
            print("\n(" + str(datetime.datetime.now()) + ")  " + blob_file)

            # Open the file and load its content
            with open(data_folder+blob_file, 'r') as file:
                file_content = file.read()
                json_data = json.loads(file_content)

            # Get the total number of documents in the file
            total_number_of_documents = len(json_data)

            if total_number_of_documents >= 0:
                # Get the collection name from the file name
                collection_name = blob_file[:blob_file.find(".json")]
  
                # Get the database and collection
                db = client[cosmos_db_mongodb_database]
                collection = db[collection_name]
                current_doc_idx = 0

                operations = []

                index_list = []

                # Iterate over each document in the JSON data
                for doc in json_data:
                    current_doc_idx = current_doc_idx + 1

                    # Generate embeddings for the document based on the collection type
                    if collection_name == "customers" and process_customers_vector:
                        doc = Customers.generateCustomerEmbedding(doc,embeddings_deployment,AzureOpenAIClient)

                    elif collection_name == "products" and process_products_vector:
                        doc = Products.generateProductEmbedding(doc,embeddings_deployment,AzureOpenAIClient)

                    elif collection_name == "salesOrders" and process_sales_orders_vector:
                        doc = SalesOrders.generateSalesOrderEmbedding(doc,embeddings_deployment,AzureOpenAIClient)
                    
                    # Print progress for every 100 documents processed
                    if current_doc_idx % 100 == 0 and ((process_customers_vector and collection_name == "customers") or (process_products_vector and collection_name == "products") or (process_sales_orders_vector and collection_name == "salesOrders")):
                        print(f"\t{current_doc_idx} out of {total_number_of_documents} docs vectorized.")

                    # Prepare the update operation for the document
                    operations.append(UpdateOne({"_id": doc["_id"]},{"$set": doc}, upsert=True))

                    # Write to the collection in batches
                    if (len(operations) == batch_size):
                        print(f"\tWriting collection {collection_name}, batch size {batch_size}, batch {batch_number}, number of documents processed so far {current_doc_idx}.")
                        collection.bulk_write(operations,ordered=False)
                        operations = []
                        batch_number = batch_number + 1
                    
                # Print the total number of documents vectorized
                if (process_customers_vector and collection_name == "customers") or (process_products_vector and collection_name == "products") or (process_sales_orders_vector and collection_name == "salesOrders"):
                    print(f"\t{total_number_of_documents} out of {total_number_of_documents} docs vectorized.")

                # Write any remaining operations to the collection
                if (len(operations) > 0):
                    print(f"\tWriting collection {collection_name}, batch size {batch_size}, batch {batch_number}, number of documents processed so far {current_doc_idx}.")
                    collection.bulk_write(operations,ordered=False)
                
                print(f"(" + str(datetime.datetime.now()) + ")  " + f"Collection {collection_name}, total number of documents processed {current_doc_idx} .\n")

                pass  # Replace this line with the lab's code

File: 04-vector-search/node.js/Blobs/loadAndVectorize.js
// Importing required modules
const fs = require('fs');
const path = require('path');
const mongodb = require('mongodb');
const { MongoClient, updateOne } = require('mongodb');

// Importing custom modules
const Customers = require('../Collections/customers');
const Products = require('../Collections/products');
const SalesOrders = require('../Collections/salesOrders');
const Indexes = require('../SearchComponents/indexes');

// Function to load and vectorize local blob data to MongoDB cluster
async function loadAndVectorizeLocalBlobDataToMongoDBCluster(client, dataFolder, cosmosDbMongoDbDatabase, batchSize, embeddingsDeployment, AzureOpenAIClient, processCustomersVector, processProductsVector, processSalesOrdersVector) {
    // Read JSON documents from the data folder
    const localBlobsFiles = fs.readdirSync(dataFolder);

    // Loop through each file in the data folder
    for (const blobFile of localBlobsFiles) {
        let batchNumber = 1;

        // Process only JSON files
        if (blobFile.includes('.json')) {
            console.log(`\n(${new Date().toISOString()})  ${blobFile}`);

            // Read the content of the file and parse it as JSON
            const fileContent = fs.readFileSync(path.join(dataFolder, blobFile), 'utf-8');
            const jsonData = JSON.parse(fileContent);

            const totalNumberOfDocuments = jsonData.length;

            // Process only if there are documents in the JSON file
            if (totalNumberOfDocuments >= 0) {
                // Get the collection name from the file name
                const collectionName = blobFile.split(".json")[0];

                // Get the database and the collection
                const db = client.db(cosmosDbMongoDbDatabase);
                const collection = db.collection(collectionName);
                let currentDocIdx = 0;

                let operations = [];

                let indexList = [];

                // Loop through each document in the JSON file
                for (let doc of jsonData) {
                    currentDocIdx++;

                    // Generate embeddings for the document based on the collection name
                    if (collectionName === "customers" && processCustomersVector) {
                        doc = await Customers.generateCustomerEmbedding(doc, embeddingsDeployment, AzureOpenAIClient);
                    } else if (collectionName === "products" && processProductsVector) {
                        doc = await Products.generateProductEmbedding(doc, embeddingsDeployment, AzureOpenAIClient);
                    } else if (collectionName === "salesOrders" && processSalesOrdersVector) {
                        doc = await SalesOrders.generateSalesOrderEmbedding(doc, embeddingsDeployment, AzureOpenAIClient);
                    }

                    // Log the progress for every 100 documents processed
                    if (currentDocIdx % 100 === 0 && ((processCustomersVector && collectionName === "customers") || (processProductsVector && collectionName === "products") || (processSalesOrdersVector && collectionName === "salesOrders"))) {
                        console.log(`\t${currentDocIdx} out of ${totalNumberOfDocuments} docs vectorized.`);
                    }

                    // Prepare the update operation for the document
                    operations.push({
                        updateOne: {
                            filter: { "_id": doc["_id"] },
                            update: { "$set": doc },
                            upsert: true
                        }
                    });

                    // Write the operations to the database in batches
                    if (operations.length === batchSize) {
                        console.log(`\tWriting collection ${collectionName}, batch size ${batchSize}, batch ${batchNumber}, number of documents processed so far ${currentDocIdx}.`);
                        await collection.bulkWrite(operations, { ordered: false });
                        operations = [];
                        batchNumber++;
                    }
                }

                // Log the completion of vectorization
                if ((processCustomersVector && collectionName === "customers") || (processProductsVector && collectionName === "products") || (processSalesOrdersVector && collectionName === "salesOrders")) {
                    console.log(`\t${totalNumberOfDocuments} out of ${totalNumberOfDocuments} docs vectorized.`);
                }

                // Write any remaining operations to the database
                if (operations.length > 0) {
                    console.log(`\tWriting collection ${collectionName}, batch size ${batchSize}, batch ${batchNumber}, number of documents processed so far ${currentDocIdx}.`);
                    await collection.bulkWrite(operations, { ordered: false });
                }

                console.log(`(${new Date().toISOString()})  Collection ${collectionName}, total number of documents processed ${currentDocIdx} .\n`);

                // Replace this line with the lab's code
                
            }
        }
    }
}

// Export the function
module.exports.loadAndVectorizeLocalBlobDataToMongoDBCluster = loadAndVectorizeLocalBlobDataToMongoDBCluster;